### PR TITLE
#2545 symptomatic

### PR DESCRIPTION
**Pull request by mouloud-mala**
#2545

### CODE REVIEW

File: sormas-api/src/main/java/de/symeda/sormas/api/symptoms/SymptomsHelper.java [hunk 1]
```diff
@@ -195,7 +196,14 @@ public static void updateIsSymptomatic(SymptomsDto dto) {
 			throw new RuntimeException(e);
 		}
 
-		dto.setSymptomatic(false);
+		if (dto.getSymptomatic() == YesNoUnknown.UNKNOWN) {
```
Comment (MateStrysewske) [inline sormas-api/src/main/java/de/symeda/sormas/api/symptoms/SymptomsHelper.java:199]: This code block doesn't really make sense; all you're doing is setting the same value that was already set before. Should be removed.

File: sormas-api/src/main/java/de/symeda/sormas/api/symptoms/SymptomsHelper.java [hunk 2]
```diff
@@ -226,13 +234,13 @@ public static void updateSymptoms(SymptomsDto sourceSymptoms, SymptomsDto target
 					if (result == SymptomState.YES) {
 						pd.getWriteMethod().invoke(targetSymptoms, result);
 					}
-				} else if (pd.getReadMethod().getReturnType() == Boolean.class) {
+				} else if (pd.getReadMethod().getReturnType() == YesNoUnknown.class) {
```
Comment (MateStrysewske) [inline sormas-api/src/main/java/de/symeda/sormas/api/symptoms/SymptomsHelper.java:237]: You're breaking existing logic here. Please revert this change. All you have to do is to remove the special condition for SymptomsDto.SYMPTOMATIC.

File: sormas-backend/src/main/java/de/symeda/sormas/backend/caze/CaseFacadeEjb.java [hunk 3]
```diff
@@ -870,7 +870,11 @@ public long count(CaseCriteria caseCriteria) {
 						pseudonymizer.pseudonymizeDto(SymptomsDto.class, visitSymptoms, inJurisdiction, null);
 
 						exportDto.setLastCooperativeVisitSymptoms(SymptomsHelper.buildSymptomsHumanString(visitSymptoms, true, userLanguage));
-						exportDto.setLastCooperativeVisitSymptomatic(visitSymptoms.getSymptomatic() ? YesNoUnknown.YES : YesNoUnknown.NO);
+						if (visitSymptoms.getSymptomatic() == YesNoUnknown.YES)
```
Comment (MateStrysewske) [inline sormas-backend/src/main/java/de/symeda/sormas/backend/caze/CaseFacadeEjb.java:873]: Always use curly brackets please, even for one-line statements

File: sormas-ui/src/main/java/de/symeda/sormas/ui/symptoms/SymptomsForm.java [hunk 4]
```diff
@@ -104,8 +105,9 @@
 					loc(SIGNS_AND_SYMPTOMS_HEADING_LOC) +
 					fluidRowCss(VSPACE_3,
 							//XXX #1620 fluidColumnLoc?
-							fluidColumn(8, 0, loc(SYMPTOMS_HINT_LOC))) +
+							fluidColumn(8, 0, loc(SYMPTOMS_HINT_LOC)) +
 					fluidRow(fluidColumn(6,6, locCss(CssStyles.ALIGN_RIGHT,BUTTONS_LOC)))+
+					fluidRowLocs(SYMPTOMATIC, ONSET_DATE) +
```
Comment (MateStrysewske) [inline sormas-ui/src/main/java/de/symeda/sormas/ui/symptoms/SymptomsForm.java:110]: Onset date and onset symptom shouldn't be split because they're related to each other. Please move onset symptom to the top as well.

File: sormas-ui/src/main/java/de/symeda/sormas/ui/visit/VisitGrid.java [hunk 5]
```diff
@@ -69,7 +70,7 @@ public VisitGrid(VisitCriteria criteria) {
 
 		((Column<VisitIndexDto, Date>) getColumn(VisitIndexDto.VISIT_DATE_TIME))
 			.setRenderer(new DateRenderer(DateHelper.getLocalDateTimeFormat(I18nProperties.getUserLanguage())));
-		((Column<VisitIndexDto, String>) getColumn(VisitIndexDto.SYMPTOMATIC)).setRenderer(new BooleanRenderer());
+		((Column<VisitIndexDto, String>) getColumn(VisitIndexDto.SYMPTOMATIC)).setRenderer(new TextRenderer());
```
Comment (MateStrysewske) [inline sormas-ui/src/main/java/de/symeda/sormas/ui/visit/VisitGrid.java:73]: You shouldn't need a renderer for this column because enums are automatically rendered according to their toString method.

File: sormas-ui/src/main/java/de/symeda/sormas/ui/symptoms/SymptomsForm.java [hunk 6]
```diff
@@ -488,6 +493,8 @@ protected void addFields() {
 
 		// Initialize lists
 
+		symptomaticFieldId = SYMPTOMATIC;
```
Comment (MateStrysewske) [inline sormas-ui/src/main/java/de/symeda/sormas/ui/symptoms/SymptomsForm.java:496]: This looks like a pretty redundant variable for me; please remove it and explicitely use SYMPTOMATIC instead

File: sormas-ui/src/main/java/de/symeda/sormas/ui/symptoms/SymptomsForm.java [hunk 7]
```diff
@@ -990,6 +1003,36 @@ private void addListenerForOnsetFields(ComboBox onsetSymptom, DateField onsetDat
 		onsetDateField.setEnabled(false); // will be updated by listener if needed
 	}
 
+	// event for the symptomatic field
+	private void addListenerForOnsetSymptomaticField(DateField onsetDateField) {
```
Comment (MateStrysewske) [inline sormas-ui/src/main/java/de/symeda/sormas/ui/symptoms/SymptomsForm.java:1007]: => addListenerForSymptomaticField (the word "onset" isn't appropriate here)

File: sormas-ui/src/main/java/de/symeda/sormas/ui/symptoms/SymptomsForm.java [hunk 8]
```diff
@@ -990,6 +1003,36 @@ private void addListenerForOnsetFields(ComboBox onsetSymptom, DateField onsetDat
...
+		Field<YesNoUnknown> sourceFieldSymptomatic = (Field<YesNoUnknown>) getFieldGroup().getField(symptomaticFieldId);
+		List<String> allPropertyIds =
+				Stream.concat(unconditionalSymptomFieldIds.stream(), conditionalBleedingSymptomFieldIds.stream()).collect(Collectors.toList());
+		allPropertyIds.add(LESIONS_THAT_ITCH);
+
+		sourceFieldSymptomatic.addValueChangeListener(event -> {
+			for (Object sourcePropertyId : allPropertyIds) {
```
Comment (MateStrysewske) [inline sormas-ui/src/main/java/de/symeda/sormas/ui/symptoms/SymptomsForm.java:1014]: The visibility status of the onset date field should only be changed once here, not for every symptom that is processed. Additionally, the onset symptom field should also be taken care of with the same logic as the onset date.

It's probably also a good idea to clear the onset date and onset symptom fields when NO or UNKNOWN are selected because if there's been a value entered before and the field is disabled, users can't remove the value anymore.

**Review by MateStrysewske — CHANGES_REQUESTED** [review]
The whole mobile app part is missing. We can't accept this pull request without the app being taken care of because it will stop working otherwise.

Also, please apply code formatting as described in the development enviroment guide to all classes you're working on.

File: sormas-ui/src/main/java/de/symeda/sormas/ui/symptoms/SymptomsForm.java (hunk 8 above)
Comment (mouloud-mala) [reply sormas-ui/src/main/java/de/symeda/sormas/ui/symptoms/SymptomsForm.java:1014]: I also add a clear on onset date when we click on clearAllButton

File: sormas-api/src/main/java/de/symeda/sormas/api/caze/GermanCaseClassificationValidator.java [hunk 9]
```diff
@@ -66,7 +67,7 @@ public static boolean isValidGermanCaseClassification(
 	private static boolean hasCoronavirusSymptom(CaseDataDto caseDataDto) {
 		final SymptomsDto symptoms = caseDataDto.getSymptoms();
 		return caseDataDto.getDisease() == Disease.CORONAVIRUS
-			&& symptoms.getSymptomatic()
+			&& symptoms.getSymptomatic() == YesNoUnknown.YES
```
Comment (MateStrysewske) [inline sormas-api/src/main/java/de/symeda/sormas/api/caze/GermanCaseClassificationValidator.java:70]: Will throw a NullPointerException if "symptomatic" is null. Instead use:
`YesNoUnknown.YES.equals(symptoms.getSymptomatic())`

File: sormas-api/src/main/java/de/symeda/sormas/api/i18n/I18nProperties.java [hunk 10]
```diff
@@ -33,6 +33,11 @@
 import java.util.PropertyResourceBundle;
 import java.util.ResourceBundle.Control;
 
+import org.apache.commons.lang3.StringUtils;
+
+import de.symeda.sormas.api.Language;
```
Comment (MateStrysewske) [inline sormas-api/src/main/java/de/symeda/sormas/api/i18n/I18nProperties.java:38]: Remove unused imports

File: sormas-api/src/main/java/de/symeda/sormas/api/symptoms/SymptomsHelper.java [hunk 11]
```diff
@@ -228,13 +235,9 @@ public static void updateSymptoms(SymptomsDto sourceSymptoms, SymptomsDto target
...
 				} else if (pd.getReadMethod().getReturnType() == Boolean.class) {
 					// Booleans are carried over when they are TRUE
-					if (pd.getName().equals(SymptomsDto.SYMPTOMATIC)) {
-						continue;
-					} else {
-						Boolean result = (Boolean) pd.getReadMethod().invoke(sourceSymptoms);
-						if (Boolean.TRUE.equals(result)) {
-							pd.getWriteMethod().invoke(targetSymptoms, result);
-						}
+					YesNoUnknown result = (YesNoUnknown) pd.getReadMethod().invoke(sourceSymptoms);
```
Comment (MateStrysewske) [inline sormas-api/src/main/java/de/symeda/sormas/api/symptoms/SymptomsHelper.java:238]: Revert this change. This doesn't make sense, you're asking whether the return type is Boolean and then process the field as if it were a YesNoUnknown field. Please just restore the former code, get rid of the if condition and only use the code in the else block.

File: sormas-backend/src/main/java/de/symeda/sormas/backend/caze/CaseFacadeEjb.java [hunk 12]
```diff
@@ -870,7 +870,13 @@ public long count(CaseCriteria caseCriteria) {
 						pseudonymizer.pseudonymizeDto(SymptomsDto.class, visitSymptoms, inJurisdiction, null);
 
 						exportDto.setLastCooperativeVisitSymptoms(SymptomsHelper.buildSymptomsHumanString(visitSymptoms, true, userLanguage));
-						exportDto.setLastCooperativeVisitSymptomatic(visitSymptoms.getSymptomatic() ? YesNoUnknown.YES : YesNoUnknown.NO);
+						if (visitSymptoms.getSymptomatic() == YesNoUnknown.YES) {
```
Comment (MateStrysewske) [inline sormas-backend/src/main/java/de/symeda/sormas/backend/caze/CaseFacadeEjb.java:873]: Why not just use `exportDto.setLastCooperativeVisitSymptomatic(visitSymptoms.getSymptomatic())`?

File: sormas-backend/src/main/java/de/symeda/sormas/backend/caze/CaseFacadeEjb.java [hunk 13]
```diff
@@ -1826,7 +1832,7 @@ private float calculateCompleteness(Case caze) {
 		if (sampleService.getSampleCountByCase(caze) > 0) {
 			completeness += 0.15f;
 		}
-		if (Boolean.TRUE.equals(caze.getSymptoms().getSymptomatic())) {
+		if ((caze.getSymptoms().getSymptomatic()) == YesNoUnknown.YES) {
```
Comment (MateStrysewske) [inline sormas-backend/src/main/java/de/symeda/sormas/backend/caze/CaseFacadeEjb.java:1835]: Will throw a NullPointerException when symptomatic is null, see above

File: sormas-backend/src/main/java/de/symeda/sormas/backend/contact/ContactFacadeEjb.java [hunk 14]
```diff
@@ -537,7 +537,11 @@ public void deleteContact(String contactUuid) {
 
 					exportContact.setLastCooperativeVisitDate(lastCooperativeVisit.getVisitDateTime());
 					exportContact.setLastCooperativeVisitSymptoms(SymptomsHelper.buildSymptomsHumanString(symptoms, true, userLanguage));
-					exportContact.setLastCooperativeVisitSymptomatic(symptoms.getSymptomatic() ? YesNoUnknown.YES : YesNoUnknown.NO);
+					if (lastCooperativeVisit.getSymptoms().getSymptomatic() == YesNoUnknown.YES) exportContact
```
Comment (MateStrysewske) [inline sormas-backend/src/main/java/de/symeda/sormas/backend/contact/ContactFacadeEjb.java:540]: See above

File: sormas-ui/src/main/java/de/symeda/sormas/ui/symptoms/SymptomsForm.java [hunk 15]
```diff
@@ -994,6 +1002,43 @@ private void addListenerForOnsetFields(ComboBox onsetSymptom, DateField onsetDat
...
+						symptom.setValue(SymptomState.NO);
+						symptom.setEnabled(false);
+					}
+				} else {
+					symptom.setEnabled(true);
+				}
+			}
+			if (sourceFieldSymptomatic.getValue() != null) {
+				if (sourceFieldSymptomatic.getValue().equals(YesNoUnknown.NO) || sourceFieldSymptomatic.getValue().equals(YesNoUnknown.UNKNOWN)) {
+					onsetDateField.setVisible(false);
```
Comment (MateStrysewske) [inline sormas-ui/src/main/java/de/symeda/sormas/ui/symptoms/SymptomsForm.java:1028]: I'd strongly recommend to use setEnabled(false) instead of setVisible(false) because otherwise users will be confused why the fields are suddenly missing

File: sormas-ui/src/main/java/de/symeda/sormas/ui/symptoms/SymptomsForm.java [hunk 16]
```diff
@@ -994,6 +1002,43 @@ private void addListenerForOnsetFields(ComboBox onsetSymptom, DateField onsetDat
...
+						symptom.setEnabled(true);
+					} else {
+						symptom.setValue(SymptomState.NO);
+						symptom.setEnabled(false);
+					}
+				} else {
+					symptom.setEnabled(true);
+				}
+			}
+			if (sourceFieldSymptomatic.getValue() != null) {
```
Comment (MateStrysewske) [inline sormas-ui/src/main/java/de/symeda/sormas/ui/symptoms/SymptomsForm.java:1026]: You can combine these two if statements because both else blocks use identical code

File: sormas-ui/src/main/java/de/symeda/sormas/ui/visit/VisitGrid.java [hunk 17]
```diff
@@ -69,7 +68,7 @@ public VisitGrid(VisitCriteria criteria) {
 
 		((Column<VisitIndexDto, Date>) getColumn(VisitIndexDto.VISIT_DATE_TIME))
 			.setRenderer(new DateRenderer(DateHelper.getLocalDateTimeFormat(I18nProperties.getUserLanguage())));
-		((Column<VisitIndexDto, String>) getColumn(VisitIndexDto.SYMPTOMATIC)).setRenderer(new BooleanRenderer());
+		getColumn(VisitIndexDto.SYMPTOMATIC);
```
Comment (MateStrysewske) [inline sormas-ui/src/main/java/de/symeda/sormas/ui/visit/VisitGrid.java:71]: Pointless line of code

File: sormas-api/src/main/java/de/symeda/sormas/api/symptoms/SymptomsHelper.java [hunk 18]
```diff
@@ -195,7 +196,13 @@ public static void updateIsSymptomatic(SymptomsDto dto) {
 			throw new RuntimeException(e);
 		}
 
-		dto.setSymptomatic(false);
+		if (dto.getSymptomatic() == YesNoUnknown.UNKNOWN) {
```
Comment (MateStrysewske) [inline sormas-api/src/main/java/de/symeda/sormas/api/symptoms/SymptomsHelper.java:199]: * See above concerning possible NullPointerExceptions
* You can write this way more efficiently like this:
```
if (!YesNoUnknown.UNKNOWN.equals(dto.getSymptomatic()) && !YesNoUnknown.NO.equals(dto.getSymptomatic())) {
  dto.setSymptomatic(null);
}
```

**Review by MateStrysewske — CHANGES_REQUESTED** [review]

File: sormas-ui/src/main/java/de/symeda/sormas/ui/contact/ContactsView.java [hunk 19]
```diff
@@ -268,25 +270,24 @@ public ContactsView() {
...
-//							caption += " (" + DateFormatHelper.getDateFormatPattern() + ")";
-//						}
-//						return caption;
-//					},
-//					createFileNameWithCurrentDate("sormas_BAG_contacts_", ".csv"),
-//					null);
-//
-//				addExportButton(bagExportResource, exportButton, exportLayout, VaadinIcons.FILE_TEXT, Captions.BAGExport, Strings.infoBAGExport);
-//			}
+			if (FacadeProvider.getConfigFacade().isConfiguredCountry(CountryHelper.COUNTRY_CODE_SWITZERLAND)
```
Comment (MateStrysewske) [inline sormas-ui/src/main/java/de/symeda/sormas/ui/contact/ContactsView.java:273]: Revert this change please, it's not associated to this issue

**Review by MateStrysewske — CHANGES_REQUESTED** [review]
I just realised that there are no changes related to the mobile app here yet; this also needs to be taken care of or the mobile app will stop working. Unfortunately that also means that you'll have to re-create the whole symptoms table thanks to SQLite being a _wonderful_ technology...

**Comment by MateStrysewske** [comment]
@mouloud-mala Actually, let's go with another approach: Please rename the current "symptomatic" field in the mobile app to something like "oldSymptomatic" and set it to `@Deprecated`. Create a new YesNoUnknown field called "symptomatic" and a new column for the symptoms table called "newSymptomatic". Add an annotation to the "symptomatic" field in Symptoms.java: `@DatabaseField(dataType = DataType.ENUM_STRING, columnName = "newSymptomatic")`.

Have a look at the "feelingIll" field for reference, we already went with this approach there.

**Comment by MateStrysewske** [comment]
Closing this pull request because it hasn't been updated since October.
